gr-digital/lib/decision_feedback_equalizer_impl.cc

```diff
@@ -0,0 +1,267 @@
+}
+
+void decision_feedback_equalizer_impl::update_decision_history(gr_complex decision)
+{
+    std::rotate(d_decision_history.begin(),
+                d_decision_history.begin() + 1,
+                d_decision_history.end());
+    d_decision_history[d_decision_history.size() - 1] = decision;
+}
+
+int decision_feedback_equalizer_impl::work(int noutput_items,
+                                           gr_vector_const_void_star& input_items,
+                                           gr_vector_void_star& output_items)
+{
+
+    auto in = static_cast<const gr_complex*>(input_items[0]);
+    auto out = static_cast<gr_complex*>(output_items[0]);
+
+    int outlen = output_items.size();
+
+    unsigned short* state = nullptr;
+    gr_complex* taps = nullptr;
+
+    if (outlen > 1)
+        taps = static_cast<gr_complex*>(output_items[1]);
+    if (outlen > 2)
+        state = static_cast<unsigned short*>(output_items[2]);
+
+    if (d_updated) {
+        gr::thread::scoped_lock guard(d_mutex);
+        d_taps = d_new_taps;
+        set_history(d_taps.size());
+        d_updated = false;
+        return 0; // history requirements may have changed.
+    }
+
+    unsigned long int nread = nitems_read(0);
+    vector<tag_t> tags;
+    vector<unsigned int> training_start_samples;
+    get_tags_in_window(
+        tags, 0, 0, noutput_items * decimation(), pmt::intern(d_training_start_tag));
+    for (const auto& tag : tags) {
+        training_start_samples.push_back(tag.offset - nread);
+    }
+
+    return equalize(in,
+                    out,
+                    noutput_items * decimation(),
+                    noutput_items,
+                    training_start_samples,
+                    true,
+                    taps,
+                    state);
+}
+
+void decision_feedback_equalizer_impl::set_taps(const vector<gr_complex>& taps)
+{
+    gr::thread::scoped_lock guard(d_mutex);
+    d_new_taps = taps;
+    d_updated = true;
+}
+
+vector<gr_complex> decision_feedback_equalizer_impl::taps() const
+{
+    gr::thread::scoped_lock guard(d_mutex);
+    return d_taps;
+}
+
+int decision_feedback_equalizer_impl::equalize(
+    const gr_complex* input_samples,
+    gr_complex* output_symbols,
+    unsigned int num_inputs,
+    unsigned int max_num_outputs,
+    std::vector<unsigned int> training_start_samples,
+    bool history_included,
+    gr_complex* taps,
+    unsigned short* state)
+{
+
+    unsigned nout = num_inputs / d_sps;
+    if (nout > max_num_outputs) {
+        nout = max_num_outputs;
+    }
+
+    const gr_complex* samples;
+    if (history_included) {
+        samples = input_samples;
+    } else {
+        vector<gr_complex> in_prepended_history(num_inputs + d_num_taps_fwd - 1);
+        std::copy(input_samples,
+                  input_samples + num_inputs,
+                  in_prepended_history.begin() + d_num_taps_fwd - 1);
+        samples = in_prepended_history.data();
```

Comment:
dangling pointer. When the vector is destroyed the line below `samples` points to unallocated data.

Reply:
Would you recommend a copy into a std::vector here?

Reply:
I'm not sure what `samples` is supposed to be, here. How long, and containing what data. My DSP algorithm skills are holding me back on understanding the intention.

E.g. `samples` if history is not included, appears to always get a history of `d_num_taps_fwd - 1` zeroes, and I don't know what that means, including if that means input samples will be dropped.

But in terms of keeping that behaviour, without the dangling pointer, maybe:

```c++
std::unique_ptr<std::vector<gr_complex>>  in_prepended_history;
if (history_samples) {
  samples = input_samples;
} else {
      in_prepended_history = boost::make_unique<std::vector<gr_complex>>(num_inputs + d_num_taps_fwd - 1);
        std::copy(input_samples,
                  input_samples + num_inputs,
                  in_prepended_history->begin() + d_num_taps_fwd - 1);
        samples = in_prepended_history->data();
}
```

But of course the copying could be skipped by creating a simple wrapper that acts like a `vector` but gives zero for anything below `N`. Then it'd be:

```c++
wrap_thing samples(input_samples);
if (!history_included) {
  samples.set_N(d_num_taps_fwd);
}
```

But maybe I missed something.

Reply:
Still getting used to smart pointers - but I like this suggestion.  

The in_prepended_history is for the case that equalize() is not being called from the work() function, and instead is a buffer of samples that doesn't relate to the previous buffer from a continuous stream.  But the filter routines work by doing dot products on the samples prior to the first new sample (hence the history).  In the case of no work function, there is no history, and I'm not maintaining it here.  So for the filter to not 1) skip samples or 2) run over invalid data, I decided to just prepad with zeros.